### PR TITLE
#2559. Add augmenting members tests. Part 2

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

**Comment by sgrekhov** [comment]
@eernstg @chloestefantsova please review

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+augment extension Ext {
+  augment static void set staticVariable(String _) {}
+//^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+}
+
+augment extension type ET {
```
Comment (eernstg) [inline LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart:64]: Here we could also add a setter that matches up with the representation variable.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t02.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// A non-writable variable declaration is augmented with a setter.
+///
+/// @description Checks that it is a compile-time error if a non-writable
+/// variable declaration is augmented with a setter. Test `late final` variables
```
Comment (eernstg) [inline LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t02.dart:10]: Oh, that should not be an error: A `late final` variable with no initializing expression _does_ have a setter, so we should be able to augment that setter.

But it's true for a `late final` variable with an initializing expression. One way ahead would be to change the expected response, another could be to add an initializing expression to all these variables.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A03_t01.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// A non-final variable is augmented with a final variable.
```
Comment (eernstg) [inline LanguageFeatures/Augmentation-libraries/augmenting_members_A03_t01.dart:7]: OK, this is one of those cases that we're discussing in https://github.com/dart-lang/language/issues/4017.

As usual, we will follow the current spec and update the test later if needed.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A04_t01.dart [hunk 4]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion A final variable can be augmented with a non-final augmenting
+/// variable, and that will not add any setter. An augmenting variable
+/// declaration only affects the initializer expression, not setters.
+///
+/// @description Checks that a final variable can be augmented with a non-final
+/// augmenting variable.
```
Comment (eernstg) [inline LanguageFeatures/Augmentation-libraries/augmenting_members_A04_t01.dart:10]: Again, I find it difficult to explain why certain things are allowed and other things are errors in this domain. But the outcome is that final/non-final makes no difference in this case at all, the only thing we're using from the augmenting declaration is the initializing expression.

**Review by eernstg — COMMENTED** [review]
I really don't know where this is going, I hope we can arrive at a clearer model in, for instance, https://github.com/dart-lang/language/issues/4017.

However, right now we'll follow the current spec wording as usual.

Please take a look at the comments. Most of them are background material. Then I'll take another look at the whole thing tomorrow and then we can take the next step.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// A non-writable variable declaration is augmented with a setter.
+///
+/// @description Checks that it is a compile-time error if a non-writable
+/// variable declaration is augmented with a setter.
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart]: > I suppose we already cover that?
No, some cases were not covered. Added, thank you! Now all of these cases are covered.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart (hunk 1 above)
Comment (sgrekhov) [reply LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart:63]: Added.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t02.dart (hunk 2 above)
Comment (sgrekhov) [reply LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t02.dart]: Indeed. Fixed by adding an initializer and also added a separate test that checks that it is not an error to augment `late final` variable with no initializer with a setter.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  final String instanceVariable = "instanceVariable";
+}
+
+class A {}
+
+extension Ext on A {
+  static final String staticVariable = "staticVariable";
+}
+
+extension type ET(String _) {
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart]: Thank you very much for the interesting and detailed explanation! I didn't think about all of this!

**Review by sgrekhov — COMMENTED** [review]
Thank you! Updated. PTAL.

**Comment by eernstg** [comment]
This one is on my TODO list, but it is full of situations that give rise to questions that haven't been clarified. I'll look into it on Monday.

**Comment by sgrekhov** [comment]
No problem, thank you!

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart [hunk 7]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion It is a compile-time error if:
+/// ...
+/// A non-writable variable declaration is augmented with a setter.
+///
+/// @description Checks that it is a compile-time error if an implicitly induced
+/// getter of a final variable declaration is augmented with a setter.
```
Comment (eernstg) [inline LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01.dart:10]: Note that there's nothing new about this, it is only targeting the situation where there is no setter to augment (in particular, there is no implicitly induced setter, but also no explicitly written setter).

So there's absolutely nothing in this error message other than the usual "can't augment something that doesn't exist" error.

File: LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart (hunk 1 above)
Comment (eernstg) [reply LanguageFeatures/Augmentation-libraries/augmenting_members_A02_t01_lib.dart:63]: Thanks!

**Review by eernstg — APPROVED** [review]
OK, let's land it!

**Pull request merged**